editor/components/inserter/menu.js

```diff
@@ -29,32 +77,206 @@ class InserterMenu extends wp.element.Component {
+		if ( null === currentBlock ) {
+			return blockTypes[ 0 ].slug;
+		}
+
+		const currentIndex = blockTypes.findIndex( ( blockType ) => currentBlock === blockType.slug );
+		const nextIndex = currentIndex + 1;
+		const highestIndex = blockTypes.length - 1;
+
+		/**
+		 * Default currently for going past the blocks is search, may need to be
+		 * revised in the future as more focusable elements are added. This
+		 * returns a null value, which currently implies that search will be set
+		 * as the next focus.
+		 */
+		if ( nextIndex > highestIndex ) {
+			return null;
+		}
+
+		// Return the slug of the next block type.
+		return blockTypes[ nextIndex ].slug;
+	}
+
+	findPrevious( currentBlock, blockTypes ) {
+		/**
+		 * null will trigger iterating back to the top of the list of block
+		 * types.
+		 */
+		if ( null === currentBlock ) {
+			return blockTypes[ 0 ].slug;
+		}
+
+		const highestIndex = blockTypes.length - 1;
+
+		// If the search bar is focused navigate to the bottom of the block list.
+		if ( 'search' === currentBlock ) {
+			return blockTypes[ highestIndex ].slug;
+		}
+
+		const currentIndex = blockTypes.findIndex( ( blockType ) => currentBlock === blockType.slug );
+		const previousIndex = currentIndex - 1;
+		const lowestIndex = 0;
+
+		/**
+		 * Default currently for going past the blocks is search, may need to be
+		 * revised in the future as more focusable elements are added. This
+		 * returns a null value, which currently implies that search will be set
+		 * as the next focus.
+		 */
+		if ( previousIndex < lowestIndex ) {
+			return null;
+		}
+
+		// Return the slug of the next block type.
+		return blockTypes[ previousIndex ].slug;
+	}
+
+	focusNext( component ) {
+		const sortedByCategory = flow(
+			this.getVisibleBlocks,
+			this.sortBlocksByCategory,
+		)( component.blockTypes );
+
+		// If the block list is empty return early.
+		if ( ! sortedByCategory.length ) {
+			return;
+		}
+
+		const currentBlock = component.state.currentFocus;
+
+		const nextBlock = this.findNext( currentBlock, sortedByCategory );
+		this.changeMenuSelection( nextBlock );
+	}
+
+	focusPrevious( component ) {
+		const sortedByCategory = flow(
+			this.getVisibleBlocks,
+			this.sortBlocksByCategory,
+		)( component.blockTypes );
+		const currentBlock = component.state.currentFocus;
+
+		// If the block list is empty return early.
+		if ( ! sortedByCategory.length ) {
+			return;
+		}
+
+		const nextBlock = this.findPrevious( currentBlock, sortedByCategory );
+		this.changeMenuSelection( nextBlock );
+	}
+
+	onKeyDown( keydown ) {
+		if ( this.isNextKeydown( keydown ) ) {
+			keydown.preventDefault();
+			this.focusNext( this );
+		}
+
+		if ( this.isPreviousKeydown( keydown ) ) {
+			keydown.preventDefault();
+			this.focusPrevious( this );
+		}
+
+		/**
+		 * Left and right arrow keys need to be handled seperately so that
+		 * default cursor behavior can be handled in the search field.
+		 */
+		if ( this.isArrowRight( keydown ) ) {
```

Comment:
I'm not overly concerned about it, but I wonder if it would have really been any less clear to just do a switch statement here instead of extracting out to a number of additional instance methods. Like:

```js
switch ( event.keyCode ) {
	case 39 /* ArrowRight */:
		if ( this.state.currentFocus !== 'search' ) {
			this.focusNext();
		}
	
	// ...
}
```